File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        int mid, start = 0, end = n;
        while(start < end) {
            mid = (end - start)/2 + start;
            if(nums[mid] == target) {
                return mid;
            }
            else if(nums[start] < nums[mid]) {
                if(nums[start] <= target && target < nums[mid]) {
                    end = mid;
                }
                else {
                    start = mid + 1;
                }
            }
            else {
                if(nums[mid] < target && target < nums[start]) {
                    start = mid + 1;
                }
                else {
                    end = mid;
                }
            }
        }
        return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)
        while left < right:
            mid = (left - right)//2 + right
            print(left, right, mid)
            if target == nums[mid]:
                return mid
            elif nums[left] < nums[mid]:
                if nums[left] <= target:
                    right = mid
                else:
                    left = mid + 1
            else:
                if nums[mid] < target <= nums[left]:
                    left = mid + 1
                else:
                    right = mid
        return -1